### DISCUSSION TITLE
ZCU106 TX data wrong

### DISCUSSION THREAD

**Discussion by MengEason** [corundum/corundum#190] [Q&A] [answered]
Hi,

When i use zcu106 pcie project, i found that tx data is wrong.
pcie rc channel data, it seems link ARP frame
128'hffff0000000000120300000b402a0202
128'h000801000608788050520b9affffffff
128'h00000a01a8c0788050520b9a01000406
128'h00000000000000001401a8c000000000
![image](https://github.com/corundum/corundum/assets/125437604/25555c2a-cad6-4c0a-84ea-dbfecc80c9b1)

rx_cpl_tlp
128'h0608788050520b9affffffffffff0000
128'ha8c0788050520b9a0100040600080100
128'h000000001401a8c00000000000000a01
![image](https://github.com/corundum/corundum/assets/125437604/79581270-9812-456f-889b-98538c820012)

DMA write data to RAM, it wrote twice to 4390f
address：4390f    BE: ffffffff        data：00000608788050520b9affffffffffff_00000608788050520b9affffffffffff
address：4390e   BE: 3fffc000    data：0100a8c0788050520b9a0100040600080100a8c0788050520b9a010004060008
address：4390f    BE: c00003ff   data：0a01000000001401a8c00000000000000a01000000001401a8c0000000000000
![image](https://github.com/corundum/corundum/assets/125437604/ee6ac318-4dd6-4f27-9cdd-55e7df64c745)
DMA read data from RAM, why it read from 4350e?
address：4350e    data：00000000000000000001000c0012c07070695f0401000c00006c61636f6c0570
address：4390e    data：0a01a8c0788050520b9a01000406000801000000000000000000000000000000
address：4390f     data：0a01a8c0788050520b9a0100040600080000060878801401a8c0000000000000
![image](https://github.com/corundum/corundum/assets/125437604/190d8deb-c28a-4f7d-b497-9e554df71546)

it convert to ethernet frame, but it is wrong
0000000000000000
0100000000000000
0b9a010004060008
0a01a8c078805052
a8c0000000000000
0000060878801401
![image](https://github.com/corundum/corundum/assets/125437604/04483a5f-d62c-462e-9c0e-63a422a4380b)

Please help check it, It seems that there is an issue with the read and write of RAM. BTW, have you been test it on ZCU106 successfully? Thanks!

**Comment by alexforencich** [comment]
Well, it worked the last time I tested it on the ZCU106.  What git revision are you using?

The first write to the RAM seems odd:

address：4390f BE: ffffffff data：00000608788050520b9affffffffffff_00000608788050520b9affffffffffff

I don't know why that was written to both segments with a BE of all FF, and the address matching the 3rd write.  Perhaps there is a problem with the write generation logic.  Also, that address is actually two addresses (one per segment), 4390f is actually (0x10f, 0x10e) and similarly 0x4350e is (0x10e, 0x10d).  So that 0x4350e with valid bits 0b01 is actually a read from 0x10e in segment 0 only, with no operation on segment 1.

**Reply by MengEason** [reply]
Hi Alex,

I use the latest git revision.

**Reply by MengEason** [reply]
meanwhile, we found that desc_block_size is different in rx and tx in mqnic_netdev.c:
ret = mqnic_open_rx_ring(q, priv, cq, priv->rx_ring_size, 1);
ret = mqnic_open_tx_ring(q, priv, cq, priv->tx_ring_size, desc_block_size);
If this configuration is used, rx will request 4 DWORDs, while tx will request 16 DWORDs
![image](https://github.com/corundum/corundum/assets/125437604/250dc518-9fe1-452f-a9c4-99fbcd4d33b7)
rq
![image](https://github.com/corundum/corundum/assets/125437604/0870466f-2de8-42f5-8c9d-1a63338b92c0)
rc
data: 00000000000000000300001040400000  keep: f
data: 00000000000000016e8798020000002a  keep: f
data: 00000000000000000000000000000000  keep: f
data: 00000000000000000000000000000000  keep: f
data: 00000000000000000000000000000000  keep: 7
![image](https://github.com/corundum/corundum/assets/125437604/4d4f476d-e8a7-4eb7-84f4-4fef2a38c0f0)
The data written to RAM will be 0
![image](https://github.com/corundum/corundum/assets/125437604/1c6ad40e-68cd-4c88-809c-35dc1c9b631d)
tx interface will request data from address 0, then tx interface will be blocked
![image](https://github.com/corundum/corundum/assets/125437604/a5621763-9a04-4a7d-8cbe-e526d90ba75f)
so we modified it like this, and tx interface will request 4 DWORDs
ret = mqnic_open_tx_ring(q, priv, cq, priv->tx_ring_size, 1);
rc
data: 00000000000000110300000440100030  keep: f
data: 000000000000000169b8a2020000002a  keep: 7
![image](https://github.com/corundum/corundum/assets/125437604/70903e87-39ac-4202-a034-4de4cc443e8c)

**Reply by alexforencich** [reply]
Don't just change stuff randomly for no reason.  The block size is set like that for a reason, to support scatter/gather on TX.  If you change it in that way, it's going to completely screw up the whole thing.

Anyway, can you try the latest development code here https://github.com/alexforencich/corundum ?

**Reply by MengEason** [reply]
As what i said, if not change "ret = mqnic_open_tx_ring(q, priv, cq, priv->tx_ring_size, desc_block_size);", tx interface will be blocked. 
What i used is from "https://github.com/corundum/corundum", is there anything different between it and the link you posted? I see you have updated something in  "https://github.com/alexforencich/corundum".  I will try it.

**Reply by MengEason** [reply]
I used the latest development code, but it's still the same as before. Is there a difference in the code between 100G and 10G? Could you please check it on ZCU106 again? Thanks! Can we communicate through email? If possible, please provide your email address.

**Reply by MengEason** [reply]
BTW, it work well when we use fpga_zynqmp project just now.

**Reply by alexforencich** [reply]
Strange.  The board-level testbench passes, so it's unclear what the problem is.  I will build and test the design on my ZCU106 to verify.  Also, if you're adding ILAs, can you probe the descriptor inputs to the PCIe DMA engine?  I can try testing the exact request in isolation to see if I can recreate the bug in sim.  But I need to know the address and length of the problematic transfer, in addition to how the RC is responding (seems like 1 completion TLP?)

**Reply by MengEason** [reply]
[tx_block_size_4.zip](https://github.com/corundum/corundum/files/15268876/tx_block_size_4.zip)
I saved ila data, hope it can be helpful, you can check it in vivado. open hardware manager, then File-->Import-->Import ILA DATA

**Reply by alexforencich** [reply]
That's perfect, I can just unzip it and open the VCD in gtkwave. Much more convenient than importing it into Vivado.

**Reply by MengEason** [reply]
[tx_block_size_1.zip](https://github.com/corundum/corundum/files/15268949/tx_block_size_1.zip)
This is another ILA data, we change tx block size to 1, ret = mqnic_open_tx_ring(q, priv, cq, priv->tx_ring_size, 1);

**Reply by MengEason** [reply]
Is there any progress?

**Reply by alexforencich** [reply, marked as answer]
Try this and let me know: https://github.com/corundum/corundum/issues/191 .  What version of Vivado are you using?

**Reply by MengEason** [reply]
I use 2023.2.1, i will try it, should i change tx block size to 1?

**Reply by MengEason** [reply]
Great, it runs normally, and the performance of the iperf test on both channels is very excellent！
![image](https://github.com/corundum/corundum/assets/125437604/892b6ceb-3e9b-41f9-98a3-a1f8e565dd38)
